src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs

```diff
@@ -511,18 +526,21 @@ IEnumerable<string> CollectSetOutputs(string setName)
                     log.Verbose(nameof(Resources.FUTD_AddingTypedOutputsInSet_2), UpToDateCheckBuilt.SchemaName, setName);
                     log.Indent++;
 
-                    foreach ((string kind, ImmutableArray<string> items) in upToDateCheckBuiltItems)
+                    foreach (string kind in state.KindNames)
                     {
-                        if (ShouldIgnoreItems(kind, items))
+                        if (upToDateCheckBuiltItems.TryGetValue(kind, out ImmutableArray<string> items))
                         {
-                            continue;
-                        }
+                            if (ShouldIgnoreItems(kind, items))
+                            {
+                                continue;
+                            }
 
-                        foreach (string path in items)
-                        {
-                            string absolutePath = _configuredProject.UnconfiguredProject.MakeRooted(path);
-                            log.VerboseLiteral(absolutePath);
-                            yield return absolutePath;
+                            foreach (string path in items)
+                            {
+                                string absolutePath = _configuredProject.UnconfiguredProject.MakeRooted(path);
+                                log.VerboseLiteral(absolutePath);
+                                yield return absolutePath;
+                            }
```

Comment:
I'd recommend using a method for these blocks of code since it is the same thing 6 times only with a change in input collection and output type. However, the non-tuple output type is just a subset of the tuple (absolutePath), so tuple output would work in all situations. Or you just make a different named method that returns absolutePath only.

Reply:
Ideally the whole loop would be extracted, but as they're generator functions I want to avoid allocating just to do that. There may be a smaller chunk (three lines) that can be extracted at no cost.

Reply:
Just out of my own interest, what would you be allocating if this was made into the method? Unless this path is hit a hundreds of thousands or millions of times, I'd personally not consider the memory impact to outweigh the cost of maintenance and readability.

*Aside*: Hmm... I'm sure there's likely have an objective measurement of perf vs code simplicity. I wouldn't be surprised if there's tooling already to do those measurements. I'd be interested to see how it compares things, and how different that value would be for the style of design I lean toward.